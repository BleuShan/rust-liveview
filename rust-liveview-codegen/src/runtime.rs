use crate::helpers::{
    has_attr,
    DarlingResultExt,
};
use darling::FromMeta;
use proc_macro::TokenStream;
use proc_macro_error::*;
use quote::quote;
use syn::{
    AttributeArgs,
    Error,
    ItemFn,
    Path,
};

#[derive(Debug, FromMeta)]
pub(crate) struct Attribute {
    executor_entrypoint: Path,
    #[darling(default)]
    test: bool,
}

impl From<AttributeArgs> for Attribute {
    fn from(attr_args: AttributeArgs) -> Self {
        Self::from_list(&attr_args).expect_or_abort("An error occured while parsing the input.")
    }
}

impl Attribute {
    pub(crate) fn generate(self, mut item: ItemFn) -> TokenStream {
        let sig = &mut item.sig;

        if sig.asyncness.is_none() {
            abort!(Error::new_spanned(
                sig.fn_token,
                "Only async functions are supported"
            ));
        }

        let vis = &item.vis;
        let attrs = &item.attrs;
        let body = &item.block;
        let entrypoint = &self.executor_entrypoint;

        sig.asyncness = None;

        let attributes = if self.test && !has_attr(attrs, "test") {
            quote! {
                #[test]
                #(#attrs)*
            }
        } else {
            quote!(#(#attrs)*)
        };
        TokenStream::from(quote! {
            #attributes
            #vis #sig {
                #entrypoint(async move {
                    #body
                })
            }
        })
    }
}
