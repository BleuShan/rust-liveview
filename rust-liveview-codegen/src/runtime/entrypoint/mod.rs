use crate::helpers::new;
use proc_macro::TokenStream;
use proc_macro2::{
    Span,
    TokenStream as TokenStream2,
};
use proc_macro_error::*;
use quote::{
    quote,
    ToTokens,
};
use syn::{
    parse::{
        Parse,
        ParseStream,
        Result,
    },
    punctuated::Pair,
    spanned::Spanned,
    Error,
    FnArg,
    ItemFn,
    Path,
    Token,
};

mod main;
mod test;

pub(crate) use main::{
    MainEntryPoint,
    MainEntryPointArgs,
};
pub(crate) use test::{
    TestEntryPoint,
    TestEntryPointArgs,
};

#[inline]
pub(crate) fn set_fn_dummy(item: &ItemFn) {
    let sig = &item.sig;
    let attrs = &item.attrs;
    let ident = &sig.ident;
    let inputs = &sig.inputs;
    let vis = &item.vis;
    set_dummy(quote! {
        #(#attrs)*
        #vis fn #ident(#inputs) {}
    });
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
pub(crate) enum Executor {
    AsyncStd,
}

impl Parse for Executor {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let path: Path = input
            .parse()
            .map_err(|e| Error::new(e.span(), "Missing executor configuration."))?;
        if path.is_ident("async_std") {
            Ok(Self::AsyncStd)
        } else {
            Err(Error::new_spanned(path, "Unknown executor configuration."))
        }
    }
}

impl ToTokens for Executor {
    fn to_tokens(&self, tokens: &mut TokenStream2) {
        tokens.extend(match self {
            Executor::AsyncStd => {
                quote! {
                    runtime::task::block_on
                }
            }
        });
    }
}
