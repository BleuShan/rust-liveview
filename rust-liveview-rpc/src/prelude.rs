pub use rust_liveview_common::{
    bytes::{
        Buf,
        BufMut,
        Bytes,
        BytesMut,
    },
    http::{
        self,
        Request,
        Response,
    },
};
pub use tower_service::Service;
