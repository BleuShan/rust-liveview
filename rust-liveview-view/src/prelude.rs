//! Prelude

pub use crate::{
    element::Element,
    html,
    render::{
        Render,
        RenderContext,
        Renderer,
    },
    result::Result,
    text::{
        escape,
        TextNode,
    },
};
