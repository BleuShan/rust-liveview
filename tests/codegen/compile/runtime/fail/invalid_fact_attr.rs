#![allow(unused_imports)]

use fluid::prelude::*;
use rust_liveview::runtime::{
    self,
    task,
};

fn main() {}

#[runtime::test(async_std)]
#[fact]
async fn runtime_attribute_test_should_add_the_test_attribute() {}
