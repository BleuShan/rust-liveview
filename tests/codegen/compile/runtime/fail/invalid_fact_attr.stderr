error: Second fact attribute is supplied.
  --> $DIR/invalid_fact_attr.rs:12:1
   |
12 | #[fact]
   | ^^^^^^^
   |
help: Consider adding fact to:
  --> $DIR/invalid_fact_attr.rs:11:1
   |
11 | #[runtime::test(async_std)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
